File: CSharp/SpecStudioExamplesTestsCSharp/common/InputControlValuesString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class InputControlValuesString
    {
        public string frame;
        public string roll;
        public string remaining;

        public InputControlValuesString(string frame, string roll, string remaining)
        {
            this.frame = frame;
            this.roll = roll;
            this.remaining = remaining;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static InputControlValuesString FromText(string text)
        {
            var parts = Tokens.Require(text, 3, "InputControlValues");
            return new InputControlValuesString(parts[0], parts[1], parts[2]);
        }

        public InputControlValuesString(string text)
        {
            var parsed = FromText(text);
            this.frame = parsed.frame;
            this.roll = parsed.roll;
            this.remaining = parsed.remaining;
        }

        public InputControlValuesTyped ToInputControlValuesTyped()
        {
            return new InputControlValuesTyped(
                int.Parse(this.frame),
                new Pins(this.roll),
                new Pins(this.remaining)
            );
        }

        public override string ToString()
        {
            return Tokens.Token(frame) + " " + Tokens.Token(roll) + " " + Tokens.Token(remaining);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not InputControlValuesString other) return false;
            return (DNCString == this.frame || DNCString == other.frame || object.Equals(this.frame, other.frame))
                && (DNCString == this.roll || DNCString == other.roll || object.Equals(this.roll, other.roll))
                && (DNCString == this.remaining || DNCString == other.remaining || object.Equals(this.remaining, other.remaining));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.frame);
            h.Add(this.roll);
            h.Add(this.remaining);
            return h.ToHashCode();
        }
    }
}

File: TypeScript/ExampleTests/common/InputControlValuesString.ts

import * as tokens from "./tokens.js";

export class InputControlValuesString {
  static readonly DNC_STRING = "?DNC?";

  frame: string;
  roll: string;
  remaining: string;

  constructor(frame: string = "", roll: string = "", remaining: string = "") {
    this.frame = frame;
    this.roll = roll;
    this.remaining = remaining;
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text: string): InputControlValuesString {
    const parts = tokens.require_(text, 3, "InputControlValues");
    return new InputControlValuesString(
      parts[0],
      parts[1],
      parts[2]
    );
  }

  static fromList(values: Iterable<string>): InputControlValuesString {
    const v = Array.from(values);
    const r = new InputControlValuesString();
    r.frame = v[0] ?? "";
    r.roll = v[1] ?? "";
    r.remaining = v[2] ?? "";
    return r;
  }

  toString(): string {
    return tokens.token(this.frame) + " " + tokens.token(this.roll) + " " + tokens.token(this.remaining);
  }

  equals(other: InputControlValuesString): boolean {
    return (this.frame === InputControlValuesString.DNC_STRING || other.frame === InputControlValuesString.DNC_STRING || this.frame === other.frame)
      && (this.roll === InputControlValuesString.DNC_STRING || other.roll === InputControlValuesString.DNC_STRING || this.roll === other.roll)
      && (this.remaining === InputControlValuesString.DNC_STRING || other.remaining === InputControlValuesString.DNC_STRING || this.remaining === other.remaining);
  }

  static equalLists(a: InputControlValuesString[], b: InputControlValuesString[]): boolean {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]!));
  }
}
